File: app/src/main/java/com/denzo/wakil/Database/HotelRepository.java
package com.denzo.wakil.Database;

import android.content.Context;
import com.denzo.wakil.Util.Hotel;
import com.denzo.wakil.Util.Reader;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class HotelRepository {
    private final BookingDao bookingDao;
    private final HousePostDao housePostDao;
    private final DraftDao draftDao;
    private final BlockedDao blockedDao;
    private final Context context;
    private final ExecutorService executor = Executors.newFixedThreadPool(4);

    public HotelRepository(Context context) {
        AppDatabase db = AppDatabase.getInstance(context);
        this.bookingDao = db.bookingDao();
        this.housePostDao = db.housePostDao();
        this.draftDao = db.draftDao();
        this.blockedDao = db.blockedDao();
        this.context = context.getApplicationContext();
    }

    public List<Hotel> getStaticHotels() {
        return Reader.getRestaurantList(context);
    }

    public List<HousePostEntity> getUserPosts() {
        return housePostDao.getAllPosts();
    }

    public List<BookingEntity> getBookings(String username) {
        return bookingDao.getBookingsByUser(username);
    }

    public void insertBooking(BookingEntity booking, Runnable onComplete) {
        executor.execute(() -> {
            bookingDao.insertBooking(booking);
            if (onComplete != null) onComplete.run();
        });
    }

    public void deleteBooking(BookingEntity booking, Runnable onComplete) {
        executor.execute(() -> {
            bookingDao.deleteBooking(booking);
            if (onComplete != null) onComplete.run();
        });
    }

    public void clearAllBookings(String username, Runnable onComplete) {
        executor.execute(() -> {
            bookingDao.clearAllBookings(username);
            if (onComplete != null) onComplete.run();
        });
    }

    public List<String> getBlockedUsers(String username) {
        return blockedDao.getBlockedUsers(username);
    }

    public List<Integer> getBlockedPosts(String username) {
        return blockedDao.getBlockedPosts(username);
    }
}

File: app/src/main/java/com/denzo/wakil/domain/repository/HotelRepository.kt
package com.denzo.wakil.domain.repository

import com.denzo.wakil.Database.BookingEntity
import com.denzo.wakil.Database.HousePostEntity
import com.denzo.wakil.Util.Hotel

interface HotelRepository {
    fun getStaticHotels(): List<Hotel>
    fun getUserPosts(): List<HousePostEntity>
    fun getBookings(username: String): List<BookingEntity>
    fun insertBooking(booking: BookingEntity, onComplete: (() -> Unit)?)
    fun deleteBooking(booking: BookingEntity, onComplete: (() -> Unit)?)
    fun clearAllBookings(username: String, onComplete: (() -> Unit)?)
    fun getBlockedUsers(username: String): List<String>
    fun getBlockedPosts(username: String): List<Int>
}
